### ISSUE TITLE
Additional fields and mappings in Elasticsearch (NHS Primary Care Domain Refsets)

### ISSUE THREAD

**Issue by evergreen-lee-campbell** [IHTSDO/snowstorm#390]
Hi!

I was wondering if the Snowstorm service would cope with the following:

- directly modifying documents in Elasticsearch to add a custom field thereto (in particular, some of this information: https://digital.nhs.uk/data-and-information/data-collections-and-data-sets/data-collections/quality-and-outcomes-framework-qof/quality-and-outcome-framework-qof-business-rules/primary-care-domain-reference-set-portal)
- adding a mapping to an existing field (i.e. https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html#add-multi-fields-existing-field-ex)

I'm particularly concerned about these either breaking the service completely, or the new fields/mappings being lost during DELTA imports. Advice very gratefully received!

**Comment by kaicode** [comment]
Hi. I think you could add additional fields without breaking the service but I don't think the additional information would be presented via the API. And you are right that the new field data would be lost if the component was updated in a new release. The best option to get this information into Snowstorm and accessible via the API would probably be to transform the data into some sort of RF2 reference set and import it in the normal way.

Could you provide an example use case to show what you are trying to achieve? I downloaded the linked Primary Care Domain Reference Set data files but I wasn't too sure what I was looking at. It looks like those data files refer to some other reference sets in the `PCD Refset ID` column. I wonder if the same data is available in RF2 elsewhere?

**Comment by evergreen-lee-campbell** [comment]
Hi Kai. I did a bit of testing on a single document to being with, and yes, I was able to add additional fields without breaking the service, but you're right: they're not presented by the API. As such, I've resorted to creating a lightweight copy of the description index, and appending the `Service_and_Ruleset` information to that, and querying it via a separate microservice. I already have a scheduled job to automatically import the DELTAs from NHS TRUD, so I'll extend that job to update my copied description index upon its creation. It seemed the safest option and meant that I was able to implement my new field mapping without having to reindex "description" itself and potentially bring down Snowstorm. 

I would wager that the `PCD Refset ID` column is indeed available elsewhere in RF2, so if I require that level of detail in future, I'll add it. It hadn't occurred to me to look, so thank you! For now, I can cope with the descriptions that live in `Service_and_Ruleset`, for my purposes.
